### PR TITLE
Add mail server and client test cases

### PR DESCRIPTION
**Pull request by QingmingSu**
Run postfix and dovecot as mail server,
test mailx, mutt, fetchmail based on the
mail server. ssl is enabled on all cases

### CODE REVIEW

**Comment by QingmingSu** [comment]
Nothing added to main.pm, the test cases are loaded as multi-machine tests via APPTESTS var.

Validation test runs for SLE12-SP3:
[mail_server]
http://10.67.18.143/tests/353
[mail_client]
http://10.67.18.143/tests/354

Validation test runs for openSUSE-Tumbleweed:
[mail_server]
http://10.67.18.143/tests/349
[mail_client]
http://10.67.18.143/tests/350

**Comment by QingmingSu** [comment]
Needle merge request for opensuse:
https://github.com/os-autoinst/os-autoinst-needles-opensuse/pull/264

Needle merge request for SLE:
https://gitlab.suse.de/openqa/os-autoinst-needles-sles/merge_requests/480

File: lib/mailtest.pm [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+);
+
+our $mail_server_name = undef;
+our $mail_server_ip   = undef;
+our $mail_client_ip   = undef;
+
+# Disable postfix dns lookup if mail server isn't FQDN
+sub postfix_dns_lookup_off {
+    script_run "sed -i 's/^\\(disable_dns_lookups\\) = .*/\\1 = yes/' /etc/postfix/main.cf";
+    script_run "systemctl restart postfix";
```
Comment (rwx788) [inline lib/mailtest.pm]: Please, sue systemctl wrapper from utils: https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/utils.pm#L147

File: lib/mailtest.pm [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+# Update postfix config file to avoid some known issues
+sub postfix_config_update {
+    #https://bugzilla.suse.com/show_bug.cgi?id=1057349
+    if (is_tumbleweed) {
+        script_run "sed -i '/^smtpd_sasl_path =\$/s/^/#/' /etc/postfix/main.cf";
+        script_run "sed -i '/^smtpd_sasl_type =\$/s/^/#/' /etc/postfix/main.cf";
+    }
+    script_run "systemctl restart postfix";
```
Comment (rwx788) [inline lib/mailtest.pm]: Please, sue systemctl wrapper from utils: https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/utils.pm#L147

File: tests/console/dovecot_ssl.pm [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    assert_script_run "curl " . data_url('dovecot/10-mail.conf') . " -o $dovecot_conf_dir/10-mail.conf";
+    assert_script_run "curl " . data_url('dovecot/10-ssl.conf') . " -o $dovecot_conf_dir/10-ssl.conf";
+
+    # Make sure user dovecot could read the certificate and key files:
+    assert_script_run "chmod 640 $dovecot_ssl_dir/*";
+    assert_script_run "chgrp dovecot $dovecot_ssl_dir/*";
+
+    # Start dovecot service
+    assert_script_run "systemctl restart dovecot.service";
+    assert_script_run "systemctl is-active dovecot.service";
```
Comment (rwx788) [inline tests/console/dovecot_ssl.pm]: Please, sue systemctl wrapper from utils: https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/utils.pm#L147

File: tests/console/dovecot_ssl.pm [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    # Dovecot will generate SSL parameters at first start up, it takes minutes for long parameters (i.e. 2048)
+    script_run "while ! (systemctl -l --no-pager status dovecot.service | grep -q 'SSL parameters regeneration completed'); do sleep 30; done", 900;
+
+    # Print service status for debugging
+    script_run "systemctl -l --no-pager status dovecot.service";
```
Comment (rwx788) [inline tests/console/dovecot_ssl.pm]: Please, sue systemctl wrapper from utils: https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/utils.pm#L147
Also, for debugging it makes sense to output it to serial device, as we then have these in one file and you don't need to find screenshot with executed command.

File: tests/console/postfix_tls.pm [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    assert_script_run "curl " . data_url('openssl/mail-server-key.pem') . " -o $postfix_key";
+    assert_script_run "sed -i 's/^#tlsmgr/tlsmgr/' /etc/postfix/master.cf";
+    assert_script_run "systemctl restart saslauthd.service";
+    assert_script_run "systemctl is-active saslauthd.service";
+    assert_script_run "systemctl restart postfix.service";
+    assert_script_run "systemctl is-active postfix.service";
+
+    # Print service status for debugging
+    script_run "systemctl -l --no-pager status saslauthd.service";
+    script_run "systemctl -l --no-pager status postfix.service";
```
Comment (rwx788) [inline tests/console/postfix_tls.pm]: Same as above, use wrapper function.

Comment (QingmingSu) [reply tests/console/postfix_tls.pm]: Thanks for the comments, I will change the case accordingly.

File: tests/console/fetchmail_ssl.pm [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    assert_script_run "curl " . data_url('openssl/ca-cert.pem') . " -o $test_cacert";
+
+    # Need mailx to send testing mail
+    mailx_setup(ssl => "yes", host => $mail_server_name);
+
+    # Switch to user console
+    type_string "chown $username /dev/$serialdev\n";
+    select_console "user-console";
+
+    for my $protocol (qw/pop3 imap/) {
```
Comment (mitiao) [inline tests/console/fetchmail_ssl.pm]: try `qw(pop3 imap)`

Comment (QingmingSu) [reply tests/console/fetchmail_ssl.pm]: Thanks, hope this solves the CI check error.

**Comment by QingmingSu** [comment]
Updated the case to:
   * use systemctl wrapper
   * output debugging info to serialdev
   * replace "qw/pop3 imap/" with "qw(pop3 imap)",
   * add pkcon_quit to stop PackageKit

And then verified again on SLE:
[mail_server]
http://10.67.18.143/tests/381
[mail_client]
http://10.67.18.143/tests/382

**Comment by mitiao** [comment]
LGTM

**Comment by okurz** [comment]
Hi, have these tests been enabled on o3?

**Comment by QingmingSu** [comment]
No, nothing added to products/opensuse/main.pm, the test suites should be triggered via vars APPTESTS together with MM_CLIENT on openSUSE.

**Comment by okurz** [comment]
Yes, I got it. So this is why I asked.
Ok, checking on https://openqa.suse.de/admin/test_suites I could find APPTESTS only mentioned for the scenario "fips_full_lvm_encrypt" which does not include the modules mentioned in this PR. So I assume also on osd the tests are not yet triggered. Also, for o3 I don't even know if we have a tap worker available. Please make sure we have a progress ticket that covers the additions of these new tests until we verified on both o3 and osd. Otherwise I would be afraid that the new tests you added are never used.

**Comment by QingmingSu** [comment]
Yeah, I just created a poo to track it on sle:
https://progress.opensuse.org/issues/25302

**Pull request merged**